docs/flip/1009-optimizer-api.md

```diff
@@ -0,0 +1,471 @@
+- FLIP PR: [#1011](https://github.com/google/flax/pull/1011)
+- FLIP Issue: [#1009](https://github.com/google/flax/issues/1009)
+
+Table of contents:
+
+- [Summary]
+- [Motivation]
+- [Using Optax]
+  - [Gradient Transformations]
+  - [Optax Training Step]
+  - [Multi Optimizer]
+  - [Train State]
+- [Previous API]
+  - [Optimizer and OptimizerDef]
+  - [Previous Training Step]
+- [Update Plan]
+- [Appendix]
+  - [Setup Code]
+
+# Summary
+[Summary]: #summary
+
+This FLIP proposes to replace our current `flax.optim` API (referred to as
+[previous API] in this document) with [Optax], DeepMind's optimizer library.
+
+# Motivation
+[motivation]: #motivation
+
+Our current API (referred to as [previous API] in this document) uses a pattern
+where an `Optimizer` dataclass is created from a pytree of `target` variables
+and from an `OptimizerDef` that defines how to update optimizer state,
+hyperparameters, and target variables. This pattern is relatively complex for
+implementing a simple optimizer, while being quite verbose in the typical Linen
+train step (especially when using mutable state collections).
+
+This package `flax.optim` contains some optimizers, but that list is far from
+exhaustive and ideally we would instead use JAX optimizers from a dedicated PyPi
+package.
+
+DeepMind already has a dedicated library — [Optax] — that implements a wide
+range of interesting optimizers and provides a framework to compose new
+optimizers from reusable gradient transformations.
+
+[Optax]: https://github.com/deepmind/optax
+
+# Using Optax
+[Using Optax]: #using-optax
+
+## Gradient Transformations
+[Gradient Transformations]: #gradient-transformations
+
+While [Optax] does provide predefined optimizers (like `optax.adam`, or
+`optax.sgd` with momentum), it is really a library of *gradient transformations*
+and the idiomatic way of instantiating an optimizer is by providing a
+combination of these gradient transformations. To emulate the momentum
+optimizer from the example when using the [previous API] we would write:
+
+```python
+import optax
+
+tx = optax.chain(
+    optax.trace(decay=0.9, nesterov=False),
+    optax.scale_by_schedule(lambda step: -get_learning_rate(step)),
+)
+```
+
+Remarks:
+
+- Above gradient transformation would be equivalent with the example under
+  [Optimizer and OptimizerDef] where we define a Momentum optimizer without
+  Nesterov momentum (note that the `beta` parameter corresponds to the `decay`
+  parameter of the `optax.trace()` transformation, and the learning rate is
+  applied in a second chained transformation).
+- Note that hyper parameters like `decay` or `nesterov` only exist in the inner
+  scope of the higher order functions returning the `GradientTransformation`.
+  Such a gradient transformation is currently defined as a `NamedTuple` of the
+  `init()` and the `update()` function. In principle this pattern could be
+  extended to also store parameters, maybe a point to discuss on the [Optax]
+  repo.
```

Comment:
Did you mean "...extended to also store **hyper**parameters...", instead of "parameters"?

Reply:
Absolutely! Fixed.